### DISCUSSION TITLE
On-demand ("lazy") adding of nodes and edges?

### DISCUSSION THREAD

**Discussion by JDLH** [networkx/networkx#7179] [Q&A]
Has anyone extended the NetworkX MultiDiGraph class to do on-demand adding of edges, so that as a shortest_paths algorithm looks for a path from a source to a target node, successor edges get added to the graph only as the algorithm needs to follow those edges?

I am interested in using NetworkX to find a relative path through a Posix-like file system directory tree. I plan to model directories as nodes, each keyed by their absolute pathname, which should be unique. Then for each directory, I model the neighbours — subdirectories, the parent directory, and arbitrary other directories reached by symbolic links — as directed edges from the original directory's node to nodes for the neighbour. I plan to  find and return the shortest paths from source to target using a variant of Dijkstra's algorithm, or perhaps of A*.

I expect that the algorithm will consult fewer than 0.01% of the directories in the file system, so I would like to avoid the time and memory of adding edges from the 99.99% of nodes which the algorithm will not consult. Thus, I plan to do on-demand ("lazy") adding of edges. I would like the graph to add nodes without adding their successor edges, and add them only when a directory is a neighbour of a directory already in the graph. Then, when code attempts to traverse these nodes, only then read the file system directory corresponding to the node, and add edges to the directories accessible from that directory. Some of the neighbouring directories will already have nodes in the graph, others will not.

All the NetworkX code which I have read seems to assume that the entire graph is populated before running algorithms over it. I have not seen any signs of on-demand adding of nodes and edges. Have I missed something?

Before I re-invent this wheel, has anyone written code like this, from which I can learn?

Reading the code for the Graph, DiGraph, MultiGraph, and MultiDiGraph classes, it seems to me that I might be able to implement this by extending MultiDiGraph as follows:

1. When adding a node *u*, set `self._succ[u] = None` instead of `= self.adjlist_inner_dict_factory()` (ref [classes/digraph.py:459-467](https://github.com/networkx/networkx/blob/267e33e99be42ed717a69817ddfbbd5c20238a2e/networkx/classes/digraph.py#L459-L467))
2. Define a class extending `dict`, with a `__getitem__(*self, key*)` method which checks if the value of `self[key]` is `None`, and if so, 
a) sets `self._succ[u] = self.adjlist_inner_dict_factory()`, and 
b) calls a function to read the directory corresponding to *key* and add the corresponding edges (ref [Python docs, 3.3.7. **Emulating container types**, `__getitem__()`](https://docs.python.org/3/reference/datamodel.html?highlight=__getitem__#object.__getitem__))

4. Return this class from `adjlist_outer_dict_factory()`.
5. When an algorithm attempts to traverse edges out of a node *u*, with code like `G.edges(u)`, the graph object eventually calls `G.succ[u]`. My class's `__getitem__` gets called, sees that `self[key]` is `None`, adds the edges and neighbour nodes, then continues via the standard code.

There are tricky parts, like how the `__getitem__()` above will find the graph to which it should add the edges. Maybe a closure can accomplish that. (This is partly why I would like to learn from someone else who has solved these problems.)

Any suggestions?

**Comment by dschult** [comment]
Hello @JDLH.  
Your proposed approach sounds promising. I'll put some random notes/thoughts here that might help. YMMV...

- I don't know of people who have done this, but we had these kind of applications in mind when making the NX2.0 release. Not that we planned to make them, but that it would be reasonably easy for an experienced hacker-type to do. :)
- I think you don't need the step of putting None into the `self._succ[u]` position. It'd probably be simpler to make that logic be part of the object returned by `adjlist_inner_dict_factory`. I guess you are looking for a lazy-dict object.  That way you only need to replace `adjlist_inner_dict_factory`. The rest of the class stays as it is.
- The lazy-dict could be configured in many ways. Somehow it has to delay creating much. Get triggered to become active. And make itself a fully functioning dict after loading the data. 
- Based on my experience with lazy_imports (which might not be applicable here) instead of subclassing dict, make a shell object 
that stores enough info to know how to:
   1) pull the data from the remote source when needed (probably `u`. maybe the structure to call to get the neighbor info) 
   2) replace itself in the self._succ and self._pred dicts (both u and the "adj" dict-like to assign it to -- either the G._succ or G._pred).
- You could make the lazy-dict's `__getattr__`/`__get_attribute__` function activate the attribute lookup process, or perhaps just rely on `__getitem__` and/or `__setitem__`. I'm not sure if there are other methods that might be needed which is the beauty of replacing getattr. Once activated, the shell object should collect the neighbor info, make a dict and put the dict into self._pred[u] or self._succ[u] as appropriate, AND THEN return the dict as if it was there all the time.

Then once lazy-dict is ready, I think it is as "easy" as:
```python
class LazyDiGraph(nx.DiGraph):
    adjlist_inner_dict_factory = lazy_dict

G = LazyDiGraph()
```
I hope it is fairly straight-forward. (which does not always mean easy).
Let me know how it goes.   :)

Say Hiii! to Ducky for me. :}

**Reply by JDLH** [reply]
Ducky says "hi" back!

**Comment by JDLH** [comment]
Thank you for the reply, Dan @dschult .

> * I think you don't need the step of putting None into the `self._succ[u]` position. It'd probably be simpler to make that logic be part of the object returned by `adjlist_inner_dict_factory`. I guess you are looking for a lazy-dict object.  That way you only need to replace `adjlist_inner_dict_factory`. The rest of the class stays as it is.

I think you are right, that I can make this simpler than I first sketched out. However, I still think the point of intervention should be the class returned by `adjlist_outer_dict_factory()`. 

This class can be an extension of dict, but it contains an internal set `initialised_keys`. If a key `k` is in this set, then `__getitem__()` behaves like a normal `dict`. If `k` is not in the dict, then the object: 

1. Calls a method `initialise(k)` to do whatever initialisation is required. In my case, it would be to read file system information about the directory named by `k`, and add edges to the graph. This will have the effect of populating the adjlist_inner_dict at `outer_dict[k]` (or else there is not much point to using the lazy mechanism).
2. Adds `k` to the set `initialised_keys`.
3. Returns  the value `dict[k]`.

Similarly, `__delitem__()` should delete its arg `k` from `initialised_keys`, then delete `k` and its value from the dict. `__setitem__()` does not need to change.

There is still the tricky matter of getting a reference to the containing Graph object to this outer dict object. `adjlist_outer_dict_factory()` gets called by the graph class's `__init__()` method, but with no parameters (e.g. [graph.py:367](https://github.com/networkx/networkx/blob/main/networkx/classes/graph.py#L367), [digraph.py:354-355](https://github.com/networkx/networkx/blob/main/networkx/classes/digraph.py#L354-L355)). I suppose that I can include in my extended dict class a method to set the reference to the Graph object. In my class extending MultiDiGraph, in the `__init__()` method, I can call the superclass's init, then call this method and pass in `self` as the Graph object reference. My LazyDiGraph is then a little more complex than your suggestion, but still pretty simple.

Also, DiGraph calls the `adjlist_outer_dict_factory()` once for `self._adj` and a second time for `self._pred`. I want the lazy loading behaviour for the former, but not for the latter. I suppose that I can have my extended dict class set a default value of None for the Graph object reference, and if the value remains None, behave as a standard dict. 

The nice thing about intervening in the objects returned by `adjlist_outer_dict_factory()` over those returned by `adjlist_inner_dict_factory()` is that there are only one or two adjlist_outer_dict objects, they are created at graph initialisation time, and so they are easy to find and intervene in. The  adjlist_inner_dict objects are more numerous, and they get created as nodes get added, when it is harder to intervene.

By the way, trying to keep the adjlist_outer_dicts and adjlist_inner_dicts and node_dicts straight makes my brain hurt a little. :-) But this discussion is helping me clarify my thinking. Thank you.
